[storage] Add migration guide for storage packages (#18305)
Azure/azure-sdk-for-js: sdk/storage/storage-blob/MigrationGuide.md

> # Guide for migrating to `@azure/storage-blob` from `azure-storage`

Reviewer: Should we specify v12 here because there is v10 for the same package and people might get confused?

Author: I'm trying to add version info here. 
What do you think about: 
     Guide for migrating to `@azure/storage-blob` v12 from `azure-storage`
or
    Guide for migrating to `@azure/storage-blob` with version 12.*.* from `azure-storage`

Reviewer: My vote is for the first option

To improve the development experience across Azure services, a set of uniform [design guidelines](https://azure.github.io/azure-sdk/general_introduction.html) was created for all languages to drive a consistent experience with established API patterns for all services. A set of [TypeScript & JavaScript Guidelines](https://azure.github.io/azure-sdk/typescript_introduction.html) was also introduced to ensure that TypeScript clients have a natural and idiomatic feel with respect to the TypeScript and JavaScript ecosystems. The new `@azure/storage-blob` client library follows these guidelines.

### Cross Service SDK improvements

The modern `@azure/storage-blob` client library is also benefited from the cross-service improvements made to the Azure development experience, such as

- A unified logging and diagnostics pipeline offering a common view of the activities across each of the client libraries
- Use of promises rather than callbacks for a simplified programming experience
- Use of async iterators in paging APIs

## Important changes

### Package name and structure

> The modern client library is named `@azure/storage-blob` and was released beginning with version 10. The legacy client library is named `azure-storage` with version of 2.x.x or below.

Reviewer: Unresolving this comment thread as I dont see what the resolution is

> The legacy library `azure-storage` grouped functionality to work with multiple services in the same package such as `Blob`, `Queue`, `Files` and `Tables`. The new `@azure/storage-blob` is dedicated to `Blob` service. New generation packages are available for the other storage services as well: `@azure/data-tables`, `@azure/storage-queue` and `@azure/storage-file-share`. This provides more granular control on which dependencies to take on your project.

Author: Updated

> ### Constructing the clients

Reviewer: We are talking about constructing clients using only connection string here. Any reason to not show the differences for other scenarios like using the anonymous credential or AAD?

I would recommend adding those as well. You can have sub headers like
- Constructing clients using connection strings
-  Constructing clients using Azure Active Directory
- etc

Author: I added one more sample for blob, queue with AAD credentials and a sample for file with SAS token. 
Samples for other credentials are similar, we may not need to cover all of them.

Previously in `azure-storage`, you would use `createBlobService` which can be used to get an instance of the `BlobService` in order to perform service level operations.

```javascript
const azure = require("azure-storage");
const blobService = azure.createBlobService("<connection-string>");
```

> Now, in `@azure/storage-blob`, we need a `BlobServiceClient` for service level operations.

Reviewer: ```suggestion
Now, in `@azure/storage-blob`, we will be creating an instance of `BlobServiceClient` for service level operations.
```

Author: Updated

```javascript
const { BlobServiceClient } = require("@azure/storage-blob");
const blobService = BlobServiceClient.fromConnectionString("<connection-string>");
```

### Creating a container

> Previously in `azure-storage`, you would use a `BlobService` instance to create a container. The `createContainer` method would take a callback to execute once the blob container has been created. This forces sequential operations to be inside the callback, potentially creating a callback chain

Author: Updated

```javascript
const azure = require("azure-storage");
const blobService = azure.createBlobService("<connection-string>");

const containerName = "<container-name>";
blobService.createContainer(containerName, function() {
  console.log(`Container created`);
});
```

> With `@azure/storage-blob` you have access to all container level operations directly from the `BlobServiceClient`. Because the blob service client is not affinitized to any one container, it is ideal for scenarios where you need to create, delete, or list more than one blob container.

Author: Updated

```javascript
const { BlobServiceClient, StorageSharedKeyCredential } = require("@azure/storage-blob");
const containerName = "<container-name>";
const blobEndpoint = "https://<account-name>.blob.core.windows.net";

const blobService = new BlobServiceClient(
  blobEndpoint,
  new StorageSharedKeyCredential("<accountName>", "<accountKey>")
);

// Creates the container with `<container-name>`
const containerClient = await blobService.createContainer(containerName);
console.log(`Container created`);
```

If your intention is to work only in the context of a single container, it's also possible to create a container from the `ContainerClient`.

```javascript
const { ContainerClient, StorageSharedKeyCredential } = require("@azure/storage-blob");
const containerUrl = "https://<account-name>.blob.core.windows.net/<container-name>";

const containerClient = new ContainerClient(
  containerUrl,
  new StorageSharedKeyCredential("<accountName>", "<accountKey>")
);

// Creates the container with `<container-name>`
const response = await containerClient.create();
console.log(`Container created`);
```

### Uploading a blob to the container

> Previously in `azure-storage`, A `BlobService` instance would be used for blob operations. `BlobService` has methods for blob operations for each blob type. For example, `createBlockBlobFromLocalFile` would be used to upload from a local file to a block blob.

Reviewer: ```suggestion
Previously in `azure-storage`, the `BlobService` class would have methods for operations for each blob type. For example, `BlobService.createBlockBlobFromLocalFile()` would be used to upload from a local file to a block blob.
```

Author: Updated

```javascript
const azure = require("azure-storage");
const containerName = "<container-name>";
const blobName = "<blob-name>";
const filePath = "<local-file-path>";
const blobService = azure.createBlobService("<connection-string>");

blobService.createBlockBlobFromLocalFile(containerName, blobName, filePath, function() {
  console.log("Blob uploaded");
});
```

> Now in the new `@azure/storage-blob` SDK, instances of `BlockBlobClient`, `PageBlobClient` and `AppendBlobClient` would be used for blob operations. Method `uploadFile` of a `BlockBlobClient` can be used to upload from a local file to a block blob.

Reviewer: ```suggestion
Now in the new `@azure/storage-blob` SDK, we have dedicated classes `BlockBlobClient`, `PageBlobClient` and `AppendBlobClient` for each blob type. For example, `BlockBlobClient.uploadFile()` can be used to upload from a local file to a block blob.
```

Author: Updated

# Guide for migrating to `@azure/storage-blob` v12 from `azure-storage`

> This guide is intended to assist in the migration to `@azure/storage-blob` from the legacy `azure-storage` package. It will focus on side-by-side comparisons for similar operations between the two packages.

Reviewer: ```suggestion
This guide is intended to assist in the migration to version 12 of `@azure/storage-blob` from the legacy `azure-storage` package. It will focus on side-by-side comparisons for similar operations between the two packages.
```

Author: Updated

We assume that you are familiar with `azure-storage`. If you are new to the Azure Storage Blob client library for JavaScript, please refer to the [README](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/storage/storage-blob/README.md) and [samples](https://github.com/Azure/azure-sdk-for-js/tree/main/sdk/storage/storage-blob/samples) rather than this guide.

## Table of contents

- [Migration benefits](#migration-benefits)

>   - [Cross Service SDK improvements](#cross-service-sdk-improvements)

Reviewer: Typically we add two more sections under `Migration benefits` that would encourage people to upgrade
- New features
- Performance improvements

Do we have content for this? It need not be part of this PR, but it would be great to have it in a follow up. If you have ideas for this, please log an issue to track updating the migration guide

Author: I've opened an issue here to track it: https://github.com/Azure/azure-sdk-for-js/issues/19051

- [Important changes](#important-changes)
  - [Package name and structure](#package-name-and-structure)

>   - [Constructing the clients](#constructing-the-clients)

Reviewer: I would rename this to `Client constructors`. Also, we don't have a section corresponding to this header. We have two sections `Constructing the clients with connection string` and `Constructing the clients with AAD token credentials` instead. I would recommend making these h4 under `Client constructors`

Author: Updated

To improve the development experience across Azure services, a set of uniform [design guidelines](https://azure.github.io/azure-sdk/general_introduction.html) was created for all languages to drive a consistent experience with established API patterns for all services. A set of [TypeScript & JavaScript Guidelines](https://azure.github.io/azure-sdk/typescript_introduction.html) was also introduced to ensure that TypeScript clients have a natural and idiomatic feel with respect to the TypeScript and JavaScript ecosystems. The new `@azure/storage-blob` client library follows these guidelines.

### Cross Service SDK improvements

The modern `@azure/storage-blob` client library is also benefited from the cross-service improvements made to the Azure development experience, such as

- A unified logging and diagnostics pipeline offering a common view of the activities across each of the client libraries
- Use of promises rather than callbacks for a simplified programming experience
- Use of async iterators in paging APIs

## Important changes

### Package name and structure

> The modern client library is named `@azure/storage-blob` . The legacy client library is named `azure-storage`.

Reviewer: ```suggestion
The modern client library is named `@azure/storage-blob` following the [naming conventions](https://azure.github.io/azure-sdk/typescript_design.html) for the new libraries across all Azure services. The legacy client library was named `azure-storage`.
```

Author: Updated

> The legacy library `azure-storage` grouped functionality to work with multiple services in the same package such as `Blob`, `Queue`, `Files` and `Tables`. The new `@azure/storage-blob` is dedicated to `Blob` service. New generation packages are available for the other storage services as well: `@azure/data-tables`, `@azure/storage-queue`, `@azure/storage-blob-changefeed`, `@azure/storage-file-datalake` and `@azure/storage-file-share`. This provides more granular control on which dependencies to take on your project.

Reviewer: ```suggestion
The legacy library `azure-storage` grouped functionality to work with multiple services such as `Blob`, `Queue`, `Files` and `Tables` in the same package. The new `@azure/storage-blob` package is dedicated to `Blob` service. Similary, dedicated packages are available for the other storage services as well: `@azure/data-tables`, `@azure/storage-queue`, `@azure/storage-blob-changefeed`, `@azure/storage-file-datalake` and `@azure/storage-file-share`. This reduces the bundle size if you were to use any of these packages in browser applications and provides more granular control on which dependencies to take on your project.
```

Author: Updated

### Constructing the clients with connection string

> Previously in `azure-storage`, you would use `createBlobService` which can be used to get an instance of the `BlobService` in order to perform service level operations.

Reviewer: ```suggestion
Previously in `azure-storage`, you would pass the connection string to the function `createBlobService` to get an instance of the `BlobService` in order to perform operations on blobs and containers.
```

Author: Updated

```javascript
const azure = require("azure-storage");
const blobService = azure.createBlobService("<connection-string>");
```

> Now, in `@azure/storage-blob`, we will be creating an instance of `BlobServiceClient` for service level operations.

Reviewer: ```suggestion
Now, in `@azure/storage-blob`, you can pass the connection string to the static method `BlobServiceClient.fromConnectionString` to create an instance of `BlobServiceClient` to perform operations on blobs and containers.
```

Author: Updated

```javascript
const { BlobServiceClient } = require("@azure/storage-blob");
const blobService = BlobServiceClient.fromConnectionString("<connection-string>");
```

### Constructing the clients with AAD token credentials

> `azure-storage` or `@azure/storage-blob` supports to access `Blob` service with different types of credentials: anonymous, account key credentials, sas token, and AAD token credentials. This section shows samples to construct blob service clients with AAD token credentials.

Reviewer: ```suggestion
Both `azure-storage` and `@azure/storage-blob` supports to access `Blob` service by creating the client with different types of credentials: anonymous, account key credentials, sas token, and AAD token credentials. This section shows the use of AAD token credentials.
```

Author: Updated

Previously in `azure-storage`, you can invoke method `createBlobServiceWithTokenCredential` to get an instance of the `BlobService` with access token for your AAD credentials.

```javascript
const azure = require("azure-storage");
const tokenCredential = new azure.TokenCredential("<access-token>");
const blobService = azure.createBlobServiceWithTokenCredential(
  "https://<account-name>.blob.core.windows.net",
  tokenCredential
);
```

> Now, for `@azure/storage-blob`, you can use a constructor of `BlobServiceClient` which accepts token credential classes provided in `@azure/identity` package to get a `BlobServiceClient` instance with AAD token credentials. In following sample, it creates an instance of `DefaultAzureCredential` which reads credentials from environment variables `AZURE_TENANT_ID`, `AZURE_CLIENT_ID` and `AZURE_CLIENT_SECRET`, and creates a `BlobServiceClient` to consume the credential instance.

Reviewer: ```suggestion
Now, for `@azure/storage-blob`, you can pass any of the [credentials from the `@azure/identity` package](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/identity/identity/samples/AzureIdentityExamples.md) to the constructor of `BlobServiceClient` to make use of your AAD credentials. In following sample, it creates an instance of `DefaultAzureCredential` which reads credentials from environment variables `AZURE_TENANT_ID`, `AZURE_CLIENT_ID` and `AZURE_CLIENT_SECRET`, and creates a `BlobServiceClient` to consume the credential instance.
```

Author: Updated
